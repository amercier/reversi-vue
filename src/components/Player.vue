<template>
  <div class="player">
    <Stone :class="`player-stone player-stone-${color}`" :color="color" />
    <div class="player-score">{{ score }}</div>
    <div v-if="status === 'play'" class="player-status player-status--play">
      Your turn
    </div>
    <div v-if="status === 'won'" class="player-status">You win!</div>
    <div v-if="status === 'idle'" class="player-status"></div>
  </div>
</template>

<script>
import { COLORS } from '../config';
import Stone from './Stone.vue';

const STATUSES = ['idle', 'play', 'won'];

/**
 * Player information.
 */
export default {
  name: 'Player',
  components: {
    Stone,
  },
  props: {
    /**
     * Player color.
     *
     * @type {String}
     */
    color: {
      type: String,
      required: true,
      validator: value => COLORS.indexOf(value) !== -1,
    },

    /**
     * Player score.
     *
     * @type {String}
     */
    score: {
      type: Number,
      required: true,
    },

    /**
     * Player status, one of:
     * - "idle": do nothing,
     * - "play": player is playing,
     * - "won": player has won.
     *
     * @type {String}
     */
    status: {
      type: String,
      required: true,
      validator: value => STATUSES.indexOf(value) !== -1,
    },
  },
};
</script>

<style lang="scss">
@keyframes blink {
  from {
    opacity: 1;
  }
  50% {
    opacity: 0.1;
  }
  to {
    opacity: 1;
  }
}

.player {
  text-align: center;

  position: relative;
  padding-bottom: 2rem; // Room for status

  &-stone {
    width: 3rem;
    height: 3rem;
    margin: 0 auto;
  }

  &-score {
    margin: 1rem 0;
    font-size: 2rem;
  }

  &-status {
    position: absolute;
    bottom: 0;
    left: 0;
    right: 0;

    &--play {
      animation: 2s blink ease-out infinite;
    }
  }
}

// Portait mode
@media screen and (max-width: 1200px) and (orientation: portrait) {
  $status-width: 5rem;

  .player {
    display: flex;
    align-items: center;
    justify-content: center;
    padding-bottom: 0;

    &-stone {
      width: 2.5rem;
      height: 2.5rem;
      margin: 0;
    }

    &-score {
      margin: 0 1rem;
    }

    &-status {
      position: static;
      width: $status-width;
      // Shift weverything to the right a bit
      margin-right: (-0.6 * $status-width);
    }
  }
}
</style>
