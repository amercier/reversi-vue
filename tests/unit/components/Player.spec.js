import { shallowMount } from '@vue/test-utils';
import Player from '@/components/Player.vue';

describe('Player.vue', () => {
  it('renders a white player', () => {
    const player = shallowMount(Player, {
      propsData: { color: 'white', score: 10, status: 'idle' },
    });
    expect(player.element).toMatchSnapshot();
  });

  it('renders a black player', () => {
    const player = shallowMount(Player, {
      propsData: { color: 'black', score: 0, status: 'idle' },
    });
    expect(player.element).toMatchSnapshot();
  });

  it('renders a white player who has to play', () => {
    const player = shallowMount(Player, {
      propsData: { color: 'white', score: 10, status: 'play' },
    });
    expect(player.element).toMatchSnapshot();
  });

  it('renders a white player who has won', () => {
    const player = shallowMount(Player, {
      propsData: { color: 'white', score: 10, status: 'won' },
    });
    expect(player.element).toMatchSnapshot();
  });
});
