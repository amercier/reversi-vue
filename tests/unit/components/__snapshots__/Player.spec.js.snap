// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player.vue renders a black player 1`] = `
<div class="player">
  <stone-stub color="black" class="player-stone player-stone-black"></stone-stub>
  <div class="player-score">0</div>
  <!---->
  <!---->
  <div class="player-status"></div>
</div>
`;

exports[`Player.vue renders a white player 1`] = `
<div class="player">
  <stone-stub color="white" class="player-stone player-stone-white"></stone-stub>
  <div class="player-score">10</div>
  <!---->
  <!---->
  <div class="player-status"></div>
</div>
`;

exports[`Player.vue renders a white player who has to play 1`] = `
<div class="player">
  <stone-stub color="white" class="player-stone player-stone-white"></stone-stub>
  <div class="player-score">10</div>
  <div class="player-status player-status--play">
    Your turn
  </div>
  <!---->
  <!---->
</div>
`;

exports[`Player.vue renders a white player who has won 1`] = `
<div class="player">
  <stone-stub color="white" class="player-stone player-stone-white"></stone-stub>
  <div class="player-score">10</div>
  <!---->
  <div class="player-status">You win!</div>
  <!---->
</div>
`;
